src/object_store/src/object/opendal_engine/opendal_s3.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use opendal::layers::{LoggingLayer, RetryLayer};
+use opendal::services::S3;
+use opendal::Operator;
+use risingwave_common::config::ObjectStoreConfig;
+
+use super::{EngineType, OpendalObjectStore};
+use crate::object::ObjectResult;
+
+impl OpendalObjectStore {
+    /// create opendal s3 engine.
+    pub fn new_s3_engine(
+        bucket: String,
+        root: String,
+        object_store_config: ObjectStoreConfig,
+    ) -> ObjectResult<Self> {
+        // Create s3 builder.
+        let mut builder = S3::default();
+        builder.bucket(&bucket);
+        builder.root(&root);
+
+        // For AWS S3, there is no need to set an endpoint; for other S3 compatible object stores, it is necessary to set this field.
+        if let Ok(endpoint_url) = std::env::var("RW_S3_ENDPOINT") {
+            builder.endpoint(&endpoint_url);
+        }
+
+        if let Ok(region) = std::env::var("AWS_REGION") {
+            builder.region(&region);
+        } else {
+            tracing::error!("aws s3 region is not set, bucket {}", bucket);
+        }
+
+        if let Ok(access) = std::env::var("AWS_ACCESS_KEY_ID") {
+            builder.access_key_id(&access);
+        } else {
+            tracing::error!("access key id of aws s3 is not set, bucket {}", bucket);
+        }
+
+        if let Ok(secret) = std::env::var("AWS_SECRET_ACCESS_KEY") {
+            builder.secret_access_key(&secret);
+        } else {
+            tracing::error!("secret access key of aws s3 is not set, bucket {}", bucket);
+        }
+
+        if std::env::var("RW_IS_FORCE_PATH_STYLE").is_err() {
+            builder.enable_virtual_host_style();
+        }
+
+        builder.disable_config_load();
```

Comment:
`disable_config_load` is used to eliminate interference. Before I thought if a cluster is s3 compatible mode, we need do add this setting, but I didn't try this situation.

Reply:
I check the implementation in opendal. The order of the config chain is `config set explicitly by builder` > `env` > `~/.aws/config` so I think there won't be interference for s3 compatible because all configs are set explicitly by builder. However, for the standard aws s3 mode, the current implementation works even without env set.

Reply:
Make sense, so let's remove this.